### PR TITLE
Jesse/bugfix/rewrite strip rstrip

### PR DESCRIPTION
**Pull request by jjbandit**
Added a bunch of test cases while bughunting as well as rewriting strip, rstrip, and adding lstrip for good measure.

### CODE REVIEW

**Review by andychu — COMMENTED** [review]
The approach with the comparators looks good, although why not make rstrip_internal() and lstrip_internal() return an index, so they don't allocate?

Benefits:

1. The only functions that allocate will be the top level lstrip(), rstrip() and chars variants.  When we move to GC it's beneficial to have more functions that don't allocate
2. Avoiding two allocations on ' foo '.strip().  The existing code does 1.

Actually it gives me an idea ... in Zig I think functions take an implicit allocator.  So one thing that could help with the GC is to actually explicitly pass allocators through the code ... Hm.

We can talk about it more on a call

**Comment by jjbandit** [comment]
> The approach with the comparators looks good, although why not make rstrip_internal() and lstrip_internal() return an index, so they don't allocate?

Cause I was being lazy and wasn't worrying about wasting memory.

> Actually it gives me an idea ... in Zig I think functions take an implicit allocator. So one thing that could help with the GC is to actually explicitly pass allocators through the code ... Hm.

Yeah I do this a lot.  I actually usually have 2 arenas; one for scratch memory that gets rolled-back, and one for the permanent storage of whatever's getting computed. ie.:

```
thing * function_with_memory(arena *Temp, arena *Perm)
{
    BeginTemporaryMemory(Temp); // << Marks the current place on the arena

    ... allocate willy-nilly from Temp ...

    thing *Result = Allocate(thing, Perm);
    Copy(Result, TempThingWeComputed, total_size_of_temp_thing_we_computed);
    return(Result);

    // Temp gets rolled back to the place we marked at the end of the function, effectively freeing all the memory we used.
}
```

This is how the C++ parser I wrote does macro expansion, such that the result ends up in a contiguous block instead of a linked-list of a billion pointers.  A lot of use cases can get away with a single over-allocated block that gets rolled back (like `strip` and friends), but anything recursive requires the above strat.  Most of the time I use two arenas and copy so I don't have to think about if I can roll the arena back or not when I'm implementing.  When I'm done if it turns out I can use a single block, I'll change then.

Anyways, question is, does Oil have arenas that work like that?

> We can talk about it more on a call

:+1:

**Comment by jjbandit** [comment]
> > The approach with the comparators looks good, although why not make rstrip_internal() and lstrip_internal() return an index, so they don't allocate?
> 
> Cause I was being lazy and wasn't worrying about wasting memory.  That's a good suggestion though, let's talk about it on the call.
> 
> > Actually it gives me an idea ... in Zig I think functions take an implicit allocator. So one thing that could help with the GC is to actually explicitly pass allocators through the code ... Hm.
> 
> Yeah I do this a lot. I actually usually have 2 arenas; one for scratch memory that gets rolled-back, and one for the permanent storage of whatever's getting computed. ie.:
> 
> ```
> thing * function_with_memory(arena *Temp, arena *Perm)
> {
>     BeginTemporaryMemory(Temp); // << Marks the current place on the arena
> 
>     ... allocate willy-nilly from Temp ...
> 
>     thing *Result = Allocate(thing, Perm);
>     Copy(Result, TempThingWeComputed, total_size_of_temp_thing_we_computed);
>     return(Result);
> 
>     // Temp gets rolled back to the place we marked at the end of the function, effectively freeing all the memory we used.
> }
> ```
> 
> This is how the C++ parser I wrote does macro expansion, such that the result ends up in a contiguous block instead of a linked-list of a billion pointers. A lot of use cases can get away with a single over-allocated block that gets rolled back (like `strip` and friends), but anything recursive requires the above strat. Most of the time I use two arenas and copy so I don't have to think about if I can roll the arena back or not when I'm implementing. When I'm done if it turns out I can use a single block, I'll change then.
> 
> Anyways, question is, does Oil have arenas that work like that?
> 
> > We can talk about it more on a call
> 
> +1

**Comment by andychu** [comment]
Oops I realized we didn't talk enough about this on the call ...

What were the bugs here?  I'm not surprised there are bugs, since the test coverage was very low.

Also I now realize this is a prerequisite for removing the `Str0` stuff: https://github.com/oilshell/oil/issues/1221

So landing this will enable some great progress

**Comment by jjbandit** [comment]
Updated to not allocate twice in `split()` @andychu

**Review by andychu — COMMENTED** [review]
OK it's great we have all the tests now ...   We should talk about the sharing/copying issue because that can cause regressions and will affect the GC version.

----

Also the "ignored" param on the comparator seems like a smell.

I looked at Python and they chose to organize it the opposite way, so you have a SINGLE function doing strip() rstrip() lstrip()  -- turn it into `striptype` --  and then another single function doing strip(chars) rstrip(chars) lstrip(chars).

https://github.com/oilshell/oil/blob/master/Python-2.7.13/Objects/stringobject.c#L1863

I was going to suggest that we could do it with a template but I think Python's solution is pretty clean.

(Unfortunately I didn't look at it when originally writing these!)

Basically there are 6 functions and you can do it like a 2x3 or 3x2.

**Comment by andychu** [comment]
BTW I will try to pull this tomorrow and run all the tests, so you don't have to update it before then ... I'm thinking about the copy/sharing issue

File: mycpp/mylib_leaky.cc [hunk 1]
```diff
@@ -11,6 +11,12 @@
 
 Str* kEmptyString = new Str("", 0);
 
+
+Str* AllocStr(strip_result str) {
+  return new Str(str.data_, str.len_);
```
[the remark this answers is not in the document]
Comment (jjbandit) [reply mycpp/mylib_leaky.cc:16]: Correct, the old routines returned slices (as noted by the `cstring-NOTE` and the `TODO` you mentioned) so this shouldn't cause regressions.  I wrote the `AllocStr` function like that precisely _because_ the old routines returned slices;  I didn't want to alter the behavior without good reason.

That said, it's probably a good idea to return copies since the GC scheme doesn't give us a choice anyways.

**Comment by andychu** [comment]
Landed this (and also got rid of buggy copied and pasted my_runtime strip()   )!
